File: Models/Dispute.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace CreaProject.Models
{
    public class Dispute
    {
        public int DisputeId { get; set; }

        // user ID from AspNetUser table.
        public string OwnerId { get; set; }

        public string Name { get; set; }

        public string BlockId { get; set; }

        public DisputeStatus Status { get; set; }

        public DisputeResolutionMethod ResolutionMethod { get; set; }

        private const double BoundPercentageDefault = 25;
        private double? _boundPercentage;

        public double? BoundsPercentage
        {
            get => _boundPercentage * 100;
            set
            {
                if (value != null)
                    _boundPercentage = (double)value / 100;
                else
                    _boundPercentage = BoundPercentageDefault / 100;
            }
        }

        private const double RatingWeightDefault = 1.1;
        public double _ratingWeight;
        public double? RatingWeight
        {
            get => (Math.Abs(_ratingWeight - 1)) * 100;
            set
            {
                if (value != null)
                    _ratingWeight = 1 + (double)value / 100;
                else
                    _ratingWeight = RatingWeightDefault;
            }
        }

        public List<Agent> Agents { get; set; }
        public List<Good> Goods { get; set; }
        public List<AgentUtility> AgentUtilities { get; set; }
        public IEnumerable<RestrictedAssignment> RestrictedAssignments { get; set; }

        public virtual string AgentsNameList
        {
            get
            {
                return Agents != null ? string.Join(", ", Agents.Select(g => g.Name).ToList()) : "";
            }
        }

        public double AgentsShareOfEntitlement
        {
            get
            {
                var shared = 1.0;
                var numAgents = 1;
                if (Agents != null)
                {
                    numAgents = Agents.Count;
                    foreach (var agent in Agents.Where(agent => agent._shareOfEntitlement != 0))
                    {
                        shared = shared - agent._shareOfEntitlement;
                        numAgents--;
                    }
                }
                if (shared >= 0)
                    return shared / numAgents;
                else
                    return 0.0;
            }
        }

        public virtual string GoodsNameList
        {
            get
            {
                return Goods != null ? string.Join(", ", Goods.Select(g => g.Name).ToList()) : "";
            }
        }

        public double DisputeBudget
        {
            get
            {
                if (Goods == null) return 0;
                var goodsValueSum = (double)Goods.Sum(g => g.EstimatedValue);
                return goodsValueSum + (goodsValueSum * (double)_boundPercentage);

            }
        }

        [Display(Name = "Creation date")]
        [DisplayFormat(DataFormatString = "{0:dd MMMM yyyy}")]
        public DateTime? CreatedAt { get; set; }

        [Display(Name = "Update date")]
        [DisplayFormat(DataFormatString = "{0:dd MMMM yyyy}")]
        public DateTime? UpdatedAt { get; set; }
    }

    public enum DisputeResolutionMethod
    {
        Ratings,
        Bids
    }

    public enum DisputeStatus
    {
        [Display(Name = "Procedure initialization")]
        SettingUp,

        [Display(Name = "Agent's preferences")]
        Bidding,

        [Display(Name = "Solution evaluation")]
        Finalizing,

        [Display(Name = "Agreement")]
        Finalized,
        
        [Display(Name = "Rejected")]
        Rejected
    }
}

File: migrated-app/backend/app/models/dispute.py
"""
Dispute model.
"""
import enum
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Enum, Float, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from app.db.session import Base


class DisputeStatus(str, enum.Enum):
    """Dispute status workflow."""
    SETTING_UP = "setting_up"
    BIDDING = "bidding"
    FINALIZING = "finalizing"
    FINALIZED = "finalized"
    REJECTED = "rejected"


class DisputeResolutionMethod(str, enum.Enum):
    """Dispute resolution method."""
    RATINGS = "ratings"
    BIDS = "bids"


class Dispute(Base):
    """Dispute model - the central entity for fair division."""

    __tablename__ = "disputes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    owner_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    name = Column(String, nullable=False)
    status = Column(Enum(DisputeStatus), default=DisputeStatus.SETTING_UP, nullable=False, index=True)
    resolution_method = Column(Enum(DisputeResolutionMethod), default=DisputeResolutionMethod.BIDS, nullable=False)

    # Bounds percentage for bid constraints (0.0 to 1.0, stored as decimal)
    bounds_percentage = Column(Float, default=0.25, nullable=False)

    # Rating weight for rating-based resolution (default 1.1)
    rating_weight = Column(Float, default=1.1, nullable=False)

    # Blockchain anchor ID
    block_id = Column(String, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False, index=True)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    owner = relationship("User", back_populates="owned_disputes", foreign_keys=[owner_id])
    agents = relationship("Agent", back_populates="dispute", cascade="all, delete-orphan")
    goods = relationship("Good", back_populates="dispute", cascade="all, delete-orphan")
    agent_utilities = relationship("AgentUtility", back_populates="dispute", cascade="all, delete-orphan")
    restricted_assignments = relationship("RestrictedAssignment", back_populates="dispute", cascade="all, delete-orphan")

    @property
    def agents_share_of_entitlement(self) -> float:
        """Calculate the default share of entitlement for agents without custom share."""
        if not self.agents:
            return 1.0

        total_custom_share = sum(
            agent.share_of_entitlement
            for agent in self.agents
            if agent.share_of_entitlement > 0
        )
        num_agents_without_custom = sum(
            1 for agent in self.agents if agent.share_of_entitlement == 0
        )

        remaining_share = 1.0 - total_custom_share
        if remaining_share >= 0 and num_agents_without_custom > 0:
            return remaining_share / num_agents_without_custom
        return 0.0

    @property
    def dispute_budget(self) -> float:
        """Calculate total budget including bounds."""
        if not self.goods:
            return 0.0
        goods_value_sum = sum(float(good.estimated_value) for good in self.goods)
        return goods_value_sum + (goods_value_sum * self.bounds_percentage)
